### PR TITLE
Combiner

### PR DESCRIPTION
**Pull request by WindTalkerGold**
multi level combiner support prototype

### CODE REVIEW

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        public TimeSpan Interval { get; set; }
+
+        public volatile bool Running = false;
+        List<Task> reducerTasks;
+        private WordCountReducer reducer;
+        private WordCountEventQueue queue;
+
+        List<int> bucketProcessingCount;
+
+        public WordCountCombinerDemo()
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs]: remove, unncessary

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    /*
+     Arsenal.exe --MaxEventsPerBatch 100 --NumBuckets 100 
+                 --WordCountFolder D:\Coding\Git\HeavenRepo\csharp\Arsenal\Arsenal\bin\Debug\Words 
+                 --AggregateLevel 2<change this value to 1,2,3> --TotalDifferentWords 20<adjust this for number of words>
+                 --OccuranceOfSkewWords 100000<adjust this for number of words> --Interval 00:00:01
+         */
+    class WordCountCombinerDemo
+    {
+        public int MaxEventsPerBatch { get; set; }
+
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs]: remove extra whitelines

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                }
+                else if (property.PropertyType == typeof(TimeSpan))
+                {
+                    parsedValue = TimeSpan.Parse(value);
+                }
+
+                property.SetValue(this, parsedValue);
+            }
+
+            Task task = Start();
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs]: Start().Wait()

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+            await generateEventsTask;
+            Console.WriteLine("generating process finished!");
+
+            Task resetRunning = ResetRunningWhenQueuesAreEmpty();
+            await Task.WhenAll(reducerTasks);
+
+            await resetRunning;
+            Console.WriteLine("all reducing tasks finished!");
+            Console.WriteLine(string.Join(", ", bucketProcessingCount));
+
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs]: remove whiteline

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+            {
+                for (int j = 0; j < TotalDifferentWords; j++)
+                {
+                    if (j >= 5)
+                    {
+                        if (rand.NextDouble() > 0.1)
+                            continue;
+                    }
+
+                    string word = new string((char)((int)'A' + j), 10);
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs:101]: so we support at most 26 different strings?

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountReducer.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.IO;
+using MultiThread.Aggregation;
+
```
Comment (sharpeye096) [inline csharp/Arsenal/Arsenal/CombinerDemo/WordCountReducer.cs:4]: remove extra whiteline

**Review by sharpeye096 — CHANGES_REQUESTED** [review]

File: csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs (hunk 5 above)
Comment (WindTalkerGold) [reply csharp/Arsenal/Arsenal/CombinerDemo/WordCountCombinerDemo.cs:101]: seems so. Handle it later...

**Review by sharpeye096 — APPROVED** [review]

**Pull request merged**